similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java

```diff
@@ -0,0 +1,215 @@
+import org.xwiki.component.manager.ComponentManager;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+
+/**
+ * Implementation of {@link org.phenotips.data.similarity.PatientGenotypeSimilarityView} that always reveals the full
+ * patient information.
+ *
+ * @version $Id$
+ * @since 1.0M6
+ */
+public class DefaultGenotypeSimilarityView extends AbstractGenotypeSimilarityView
+{
+    /** Manager to allow access to underlying genotype data. */
+    protected static PatientGenotypeManager genotypeManager;
+
+    /** Logging helper object. */
+    private static Logger logger = LoggerFactory.getLogger(DefaultGenotypeSimilarityView.class);
+
+    /**
+     * Simple constructor passing the {@link #match matched patient}, the {@link #reference reference patient}, and the
+     * {@link #access patient access type}.
+     *
+     * @param match the matched patient, which is represented by this match
+     * @param reference the reference patient, which the match is against
+     * @param access the access type the user has to the match patient
+     * @throws IllegalArgumentException if one of the patients is {@code null}
+     */
+    public DefaultGenotypeSimilarityView(Patient match, Patient reference, AccessType access)
+        throws IllegalArgumentException
+    {
+        super(match, reference, access);
+
+        if (genotypeManager == null) {
+            genotypeManager = getGenotypeManager();
+        }
+
+        if (genotypeManager != null) {
+            // Store genotype information
+            this.matchGenotype = genotypeManager.getGenotype(this.match);
+            this.refGenotype = genotypeManager.getGenotype(this.reference);
+        }
+
+        // Match candidate genes and any genotype available
+        this.geneScores = new HashMap<String, Double>();
+        if (this.matchGenotype != null && this.refGenotype != null && this.matchGenotype.hasGenotypeData()
+            && this.refGenotype.hasGenotypeData()) {
+            matchGenes();
+        }
+    }
+
+    /**
+     * Lookup component genotype manager.
+     *
+     * @return {@link PatientGenotypeManager} component
+     */
+    private PatientGenotypeManager getGenotypeManager()
+    {
+        ComponentManager componentManager = ComponentManagerRegistry.getContextComponentManager();
+        try {
+            return componentManager.getInstance(PatientGenotypeManager.class);
+        } catch (ComponentLookupException e) {
+            logger.error("Unable to look up PatientGenotypeManager: " + e.toString());
+        }
+        return null;
+    }
+
+    /**
+     * Score genes with variants in both patients, and sets 'geneScores' and 'geneVariants' fields accordingly.
```

Comment:
The contradiction is that the code doesn't care if a gene is in both sources, it will just use `0` as the score for the missing one. Or does `getGenes` do the filtering on pairs?